packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,893 @@
+///  * [RestorationBucket], which make up the restoration data hierarchy.
+///  * [RestorationId], which identifies a value or a child uniquely within a
+///    [RestorationBucket].
+///  * [RestorationMixin], which uses [RestorationBucket]s behind the scenes
+///    to make [State] objects of [StatefulWidget]s restorable.
+class RestorationManager {
+  /// The root of the [RestorationBucket] hierarchy containing the restoration
+  /// data.
+  ///
+  /// Child buckets can be claimed from this bucket via
+  /// [RestorationBucket.claimChild]. If the [RestorationManager] has been asked
+  /// to restore the application to a previous state, these buckets will contain
+  /// the previously stored data. Otherwise the root bucket (and all children
+  /// claimed from it) will be empty.
+  ///
+  /// Like any [RestorationBucket], the root bucket informs its listeners if it
+  /// has been replaced with a new root bucket. This happens when new
+  /// restoration data has been provided to the [RestorationManager] to restore
+  /// the application to a different state. In response to the notification,
+  /// listeners must stop using the old root bucket and obtain the new one via
+  /// this getter ([rootBucket] will have been updated to return the new bucket
+  /// just before the old root bucket informs its listeners).
+  ///
+  /// The restoration data describing the current bucket hierarchy is retrieved
+  /// asynchronously from the engine the first time the root bucket is accessed
+  /// via this getter. After the data has been copied over from the engine, this
+  /// getter will return a [SynchronousFuture], that immediately resolves to the
+  /// root [RestorationBucket].
+  ///
+  /// See also:
+  ///
+  ///  * [RootRestorationScope], which makes the root bucket available in the
+  ///    [Widget] tree.
+  Future<RestorationBucket> get rootBucket {
+    if (!_isListeningForEngineUpdates) {
+      SystemChannels.restoration.setMethodCallHandler(_methodHandler);
+      _isListeningForEngineUpdates = true;
+    }
+    if (_rootBucket != null) {
+      return SynchronousFuture<RestorationBucket>(_rootBucket);
+    }
+    if (_pendingRootBucket == null) {
+      _pendingRootBucket = Completer<RestorationBucket>();
+      _getRootBucketFromEngine();
+    }
+    return _pendingRootBucket.future;
+  }
+  RestorationBucket _rootBucket;
+  Completer<RestorationBucket> _pendingRootBucket;
+  bool _isListeningForEngineUpdates = false;
+
+  Future<void> _getRootBucketFromEngine() async {
+    final Map<String, dynamic> data = await retrieveFromEngine();
+    if (_pendingRootBucket == null) {
+      // The engine was faster in sending us the data via the 'push' method on
+      // the SystemChannel.
+      return;
+    }
+    assert(_rootBucket == null);
+    _setRootBucket(data);
+  }
+
+  void _setRootBucket(Map<String, dynamic> data) {
+    _rootBucket = _createRootBucket(data);
+    _pendingRootBucket?.complete(_rootBucket);
```

Comment:
I think this is save, but maybe have an assertion above this that the completer is either null or not completed, which I think will give this a better error callstack in the case where it's already compelted.

Reply:
Added the assert. This is save because this is the only place where the completer is completed and right after its field is nulled out. So there's no chance to ever complete it again. Nevertheless, the assert is a good idea.